### PR TITLE
Fix #3359: Automatically remove colon and apostrophe from key pattern

### LINKED ISSUE #3359: key generator fails to parse colon
It seems that in 4.0 (Mac version), when I customize the key generator to [author][year][veryshorttitle], the generated keys will include the trailing colon if there is one in the title. It also includes other special characters like (`). 

I remember this wasn't the case in 3.6.

### PR DESCRIPTION
**Pull request by tobiasdiez**
Fixes #3359. Colons and apostrophes are now removed from the generated key pattern.
I also refactored the key generator, mainly converting the static methods to instance methods.

----

- [x] Change in CHANGELOG.md described
- [x] Tests created for changes
- [ ] Screenshots added (for bigger UI changes)
- [x] Manually tested changed features in running JabRef
- [ ] Check documentation status (Issue created for outdated help page at [help.jabref.org](https://github.com/JabRef/help.jabref.org/issues)?)
- [ ] If you changed the localization: Did you run `gradle localizationUpdate`?

### CODE REVIEW

**Comment by tobiasdiez** [comment]
@lenhard I manually tested most use cases, where the code changes (except OpenOffice since I don't use it). What are our current plans for the 4.1 release? When do we want to release? If there is still some time before the release I would prefer if this PR is merged before the release.

**Comment by Siedlerchr** [comment]
You may want to add the percent sign and the ampersand (&), which is problematic as it creates a problem with biber/biblatex

File: src/main/java/org/jabref/logic/bibtexkeypattern/BibtexKeyGenerator.java [hunk 1]
```diff
@@ -18,33 +22,95 @@
...
+                    newKey.append(c);
+                }
+            }
+            return newKey.toString();
+        }
+
+        StringBuilder newKey = new StringBuilder();
+        for (int i = 0; i < key.length(); i++) {
+            char c = key.charAt(i);
+            if (!Character.isWhitespace(c) && ("{}(),\\\"#~^':`".indexOf(c) == -1)) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/logic/bibtexkeypattern/BibtexKeyGenerator.java]: You should reeally extract this to a constant, same with th e one abvoe

**Comment by tobiasdiez** [comment]
Should this still go into 4.1?

**Comment by lenhard** [comment]
@tobiasdiez When you fix the checkstyle issues and conflicts then this PR can be merged.

**Pull request merged**